File: 0817-Linked List Components/0817-Linked List Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& G) {
        unordered_set<int> table;
        for (int i : G) {
            table.insert(i);
        }
        ListNode* curr = head;
        int count = 0;
        while (curr) {
            if (table.count(curr->val) && (curr->next == nullptr || !table.count(curr->next->val))) {
                ++count;
            }
            curr = curr->next;
        }
        
        return count;
    }
};

File: 0801-0900/0817-Linked List Components/0817-Linked List Components.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def numComponents(self, head: ListNode, G: List[int]) -> int:
        table = set(G)
        curr = head
        count = 0
        while curr:
            if curr.val in table and (curr.next is None or curr.next.val not in table):
                count += 1
            curr = curr.next
        return count
